File: Leetcode/Python3/2444. Count Subarrays With Fixed Bounds/solution.py
# https://leetcode.com/problems/count-subarrays-with-fixed-bounds/
# Ref: https://leetcode.com/problems/count-subarrays-with-fixed-bounds/solutions/2708099/java-c-python-sliding-window-with-explanation/?orderBy=most_votes

class Solution:
    # Ex. nums = [3 1 3 5 2 7 5], minK = 1, maxK = 5
    def countSubarrays(self, nums: List[int], minK: int, maxK: int) -> int:
        mink_index, maxk_index = None, None
        subarray_index = -1
        count = 0
        for idx, num in enumerate(nums):
            # When idx == 0, subarray_index = -1
            if not minK <= num <= maxK:  # Subarray should only have values greater than minK and less than maxK.
                subarray_index = idx
            # MinK and maxK may be the same, so inspect them separately.
            if num == minK:  # When idx == 1, mink_index = 1
                mink_index = idx
            if num == maxK:  # When idx == 3, maxk_index = 3
                maxk_index = idx
            if mink_index != None and maxk_index != None:
                # Subarray range = subarray_index + 1 ~ min(mink_index, maxk_index)
                # mink_index = 1, maxk_index = 3, subarray_index = -1, subarray = [3 1 3 5], [1 3 5]
                # subarray range = [3 1 3 5] 2 7 5
                count += max(0, min(mink_index, maxk_index) - subarray_index)
        return count

File: Leetcode/C++/2444. Count Subarrays With Fixed Bounds/solution.cpp
// https://leetcode.com/problems/count-subarrays-with-fixed-bounds/description/
// Ref: https://leetcode.com/problems/count-subarrays-with-fixed-bounds/solutions/2708099/java-c-python-sliding-window-with-explanation/?orderBy=most_votes

#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
public:
    long long countSubarrays(vector<int> &nums, int minK, int maxK)
    {
        long count = 0;
        int subArrayIndex = -1;
        int minKIndex = -1, maxKIndex = -1;
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] < minK || nums[i] > maxK)
            {
                subArrayIndex = i;
            }
            if (nums[i] == minK)
            {
                minKIndex = i;
            }
            if (nums[i] == maxK)
            {
                maxKIndex = i;
            }
            if (minKIndex != -1 || maxKIndex != -1)
            {
                count += max(0, min(minKIndex, maxKIndex) - subArrayIndex);
            }
        }
        return count;
    }
};